presto-native-execution/presto_cpp/main/expression/RowExpressionOptimizer.cpp

```diff
@@ -0,0 +1,698 @@
+          veloxToPrestoRowExpression(exprInputs[i + 1], inputWhenArgs[1]));
+
+      result.emplace_back(getWhenSpecialForm(whenArgs, i));
+    }
+  } else {
+    // The case 'expression' in simple form of conditional cannot be inferred
+    // from Velox since it could evaluate all when clauses to true or false, so
+    // we get it from the input json.
+    result.emplace_back(inputArgs[0]);
+    for (auto i = 0; i < numInputs - 1; i += 2) {
+      json::array_t whenArgs;
+      const vector_size_t argsIdx = i / 2 + 1;
+      const auto& caseValue = exprInputs[i + 1];
+      json::array_t inputWhenArgs = inputArgs[argsIdx].at("arguments");
+
+      if (exprInputs[i]->isConstant()) {
+        auto constantExpr =
+            std::dynamic_pointer_cast<const exec::ConstantExpr>(exprInputs[i]);
+        if (auto constVector =
+                constantExpr->value()->as<ConstantVector<bool>>()) {
+          if (constVector->valueAt(0)) {
+            if (result.size() == 1) {
+              // This is the first case statement that evaluates to true, so
+              // return the expression corresponding to this case.
+              return {
+                  json::array({veloxToPrestoRowExpression(
+                      caseValue, inputWhenArgs[1])}),
+                  false};
+            } else {
+              // If the case has been constant folded to false in the Velox
+              // switch expression, we do not have access to the expression
+              // inputs in Velox anymore. So we return the corresponding
+              // argument from the input switch expression.
+              result.emplace_back(inputArgs[argsIdx]);
+            }
+          } else {
+            // Skip cases that evaluate to false from the output switch
+            // expression's arguments.
+            continue;
+          }
+        } else {
+          whenArgs.emplace_back(getConstantRowExpression(constantExpr));
+        }
+      } else {
+        VELOX_USER_CHECK(!exprInputs[i]->inputs().empty());
+        const auto& matchExpr = exprInputs[i]->inputs().back();
+        whenArgs.emplace_back(
+            veloxToPrestoRowExpression(matchExpr, inputWhenArgs[0]));
+      }
+
+      whenArgs.emplace_back(
+          veloxToPrestoRowExpression(caseValue, inputWhenArgs[1]));
+      result.emplace_back(getWhenSpecialForm(whenArgs, i));
+    }
+  }
+
+  // Else clause.
+  if (numInputs % 2 != 0) {
+    result.push_back(veloxToPrestoRowExpression(
+        exprInputs[numInputs - 1], inputArgs[numArgs - 1]));
+  }
+  return {result, true};
+}
+
+json RowExpressionConverter::getSpecialForm(
+    const exec::ExprPtr& expr,
+    const json& input) {
+  json res;
+  res["@type"] = kSpecial;
+  auto form = expr->name();
+  // Presto requires the field form to be in upper case.
+  std::transform(form.begin(), form.end(), form.begin(), ::toupper);
+  res["form"] = form;
+  auto exprInputs = expr->inputs();
+  res["arguments"] = json::array();
```

Comment:
Move this line after the comment and just above the if condition that involves setting it.